src/particle.cpp

```diff
@@ -0,0 +1,681 @@
+        std::to_string(surf->id_));
+    }
+    return;
+  }
+
+  // ==========================================================================
+  // SEARCH NEIGHBOR LISTS FOR NEXT CELL
+
+#ifdef DAGMC
+  if (settings::dagmc) {
+    auto cellp = dynamic_cast<DAGCell*>(model::cells[cell_last_[0]].get());
+    // TODO: off-by-one
+    auto surfp = dynamic_cast<DAGSurface*>(model::surfaces[std::abs(surface_) - 1].get());
+    int32_t i_cell = next_cell(cellp, surfp) - 1;
+    // save material and temp
+    material_last_ = material_;
+    sqrtkT_last_ = sqrtkT_;
+    // set new cell value
+    coord_[0].cell = i_cell;
+    cell_instance_ = 0;
+    material_ = model::cells[i_cell]->material_[0];
+    sqrtkT_ = model::cells[i_cell]->sqrtkT_[0];
+    return;
+  }
+#endif
+
+  if (find_cell(this, true)) return;
+
+  // ==========================================================================
+  // COULDN'T FIND PARTICLE IN NEIGHBORING CELLS, SEARCH ALL CELLS
+
+  // Remove lower coordinate levels and assignment of surface
+  surface_ = 0;
+  n_coord_ = 1;
+  bool found = find_cell(this, false);
+
+  if (settings::run_mode != RUN_MODE_PLOTTING && (!found)) {
+    // If a cell is still not found, there are two possible causes: 1) there is
+    // a void in the model, and 2) the particle hit a surface at a tangent. If
+    // the particle is really traveling tangent to a surface, if we move it
+    // forward a tiny bit it should fix the problem.
+
+    n_coord_ = 1;
+    this->r() += TINY_BIT * this->u();
+
+    // Couldn't find next cell anywhere! This probably means there is an actual
+    // undefined region in the geometry.
+
+    if (!find_cell(this, false)) {
+      this->mark_as_lost("After particle " + std::to_string(id_) +
+        " crossed surface " + std::to_string(surf->id_) +
+        " it could not be located in any cell and it did not leak.");
+      return;
+    }
+  }
+}
+
+void
+Particle::mark_as_lost(const char* message)
+{
+  // Print warning and write lost particle file
+  warning(message);
+  write_restart();
+
+  // Increment number of lost particles
+  alive_ = false;
+  #pragma omp atomic
+  simulation::n_lost_particles += 1;
+
+  // Count the total number of simulated particles (on this processor)
+  auto n = simulation::current_batch * settings::gen_per_batch *
+    simulation::work_per_rank;
+
+  // Abort the simulation if the maximum number of lost particles has been
+  // reached
+  if (simulation::n_lost_particles >= MAX_LOST_PARTICLES &&
+      simulation::n_lost_particles >= REL_MAX_LOST_PARTICLES*n) {
+    fatal_error("Maximum number of lost particles has been reached.");
+  }
+}
+
+void
+Particle::write_restart() const
+{
+  // Dont write another restart file if in particle restart mode
+  if (settings::run_mode == RUN_MODE_PARTICLE) return;
+
+  // Set up file name
+  std::stringstream filename;
+  filename << settings::path_output << "particle_" << simulation::current_batch
+    << '_' << id_ << ".h5";
```

Comment:
Move strings into constant variables instead of hard-coding them inlined like this.